File: leetcode/graph/82. Remove Duplicates from Sorted List II/py/main.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

# Time O(N)
# Space O(N)
class Solution:
    def deleteDuplicates(self, head: ListNode) -> ListNode:
        if not head: return head

        dummy = ListNode()
        p = dummy

        while head and head.next:
          if head.val != head.next.val:
            p.next = head
            p = p.next
            head = head.next
          else:
            a = head.next

            while a and a.val == head.val:
              a = a.next    
            
            head = a
         
        p.next = head

        return dummy.next

File: leetcode/graph/82. Remove Duplicates from Sorted List II/js/main.js
// Time O(N)
// Space O(N)
const deleteDuplicates = head => {
  if (head === null) return null;

  if (head && !head.next) return new ListNode(head.val);

  let dummy = new ListNode(0);
  let a = dummy;

  while (head && head.next) {
    if (head.val !== head.next.val) {
      a.next = new ListNode(head.val);
      a = a.next;
      head = head.next;
    } else {
      let val = head.val;
      head = head.next;

      while (head && head.next && head.val === val) {
        head = head.next;
      }

      if (head.val === val) {
        head = null;
      }
    }
  }

  if (head) {
    a.next = new ListNode(head.val);
  }

  return dummy.next;
};
